sale.py

```diff
@@ -57,3 +50,31 @@ def get_shipment_state(self):
                 if any(s for s in self.shipments if s.state in ('done', 'cancelled')):
                     return 'sent'
         return super(Sale, self).get_shipment_state()
+
+    @classmethod
+    def process(cls, sales):
+        pool = Pool()
+        SaleLine = pool.get('sale.line')
+        ShipmentOut = pool.get('stock.shipment.out')
+
+        super().process(sales)
+
+        for sale in sales:
+            if (sale.shipment_state == 'sent'
+                    and sale.remaining_stock == 'manual'):
```

Comment:
I don't think that is correct. A partially delivered sale will not have shipment_state == 'sent'.

I think the right way to do it should be to check all stock moves of all lines. For each line that has two or more stock moves and at least one of them is in done state, the other one should be cancelled or cancel its shipment.

The problem with cancelling the shipment is that the line may have been added into another shipment if shipment grouping is configured.

Reply:
Remove set shipment_state to sent by code ; Control shipment state from sale core module